packages/core/src/workflow/piece-add.js

```diff
@@ -0,0 +1,128 @@
+import { Aggregator } from '@web3-storage/filecoin-client'
+import * as Server from '@ucanto/server'
+import { decode as decodePiece } from '../data/piece.js'
+
+/**
+ * @typedef {import('@web3-storage/data-segment').PieceLink} PieceLink
+ * @typedef {import('../data/types.js').Piece<PieceLink>} Piece
+ * 
+ * @typedef {object} EncodedRecord
+ * @property {string} body
+ * @property {string} id
+ * 
+ * @typedef {object} Record
+ * @property {Piece} piece
+ * @property {string} id
+ */
+
+/**
+ * Invokes aggregate/add on the queued system to go through the backlog of pieces provided by Storefront actors.
+ * Once the `aggregate/add` succeeds, the peice is queued for buffering on its way to be aggregated.
+ * When only a subset of the pieces succeed, a `batchItemFailures` is returned so that these items can be re-queued.
+ *
+ * @param {object} props
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Piece>} props.queueClient
+ * @param {import('@web3-storage/filecoin-client/types').InvocationConfig} props.invocationConfig
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} props.aggregatorServiceConnection
+ * @param {EncodedRecord[]} props.records - message encoded piece records
+ */
+export async function addPieces ({
+  queueClient,
+  invocationConfig,
+  aggregatorServiceConnection,
+  records,
+}) {
+  const decodedRecords = await Promise.all(
+    records.map(async pr => ({
+      piece: await decodePiece.message(pr.body),
+      id: pr.id
+    }))
+  )
+
+  let responses
+  try {
+    responses = await Promise.all(
+      decodedRecords.map(record => addPiece({
+        queueClient,
+        invocationConfig,
+        aggregatorServiceConnection,
+        record
+      }))
+    )
+  } catch {
+    return {
+      error: new AggregateAddFailed('failed to add pieces')
+    }
+  }
+
+  return {
+    ok: {
+      // number of successful handled pieces
+      countSuccess: responses.reduce((acc, value) => {
+        return value.ok ? acc + 1 : acc
+      }, 0),
+      // failed pieces that can be re-queued
+      batchItemFailures: responses.reduce((filtered, value) => {
+        if (value.error) {
+          filtered.push(value.id)
+        }
+
+        return filtered
+      }, /** @type {string[]} */ ([]))
+    }
+  }
```

Comment:
Typically one would represent result of such a function as `Result<Ok, Err>[]` instead `Result<{ ok: T[], error: Err[] }, Fail>`, although there might be cases where custom type might do better.


I have not seen function using this yet, but I suspect it could do filtering over array instead.

Reply:
It also looks like you just return success count, which I doubt is of any practical use, if my assumption is correct you could instead return `Result<{}, Err[]>` where you return `Ok` if there were no failures, otherwise you return the errors, because per your comment you treat partial success as an error.